web/packages/teleport/src/DesktopSession/DesktopSession.tsx

```diff
@@ -117,8 +200,100 @@ export function DesktopSession(props: State) {
     mfa,
   ]);
 
+  const tdpClientCanvasRef = useRef<TdpClientCanvasRef>(null);
+
+  useEffect(() => {
+    if (!client) {
+      return;
+    }
+    const setPointer = tdpClientCanvasRef.current?.setPointer;
+    client.addListener(TdpClientEvent.POINTER, setPointer);
+
+    return () => {
+      client.removeListener(TdpClientEvent.POINTER, setPointer);
+    };
+  }, [client]);
+
+  const onInitialTdpConnectionSucceeded = useCallback(() => {
+    setInitialTdpConnectionSucceeded(() => {
+      // TODO(gzdunek): This callback is a temporary fix for focusing the canvas.
+      // Focus the canvas once we start rendering frames.
+      // The timeout it a small hack, we should verify
+      // what is the earliest moment we can focus the canvas.
+      setTimeout(() => {
+        tdpClientCanvasRef.current?.focus();
+      }, 100);
+    });
+  }, [setInitialTdpConnectionSucceeded]);
+
+  useEffect(() => {
+    if (!client) {
+      return;
+    }
+    const renderFrame = (frame: PngFrame) => {
+      onInitialTdpConnectionSucceeded();
+      tdpClientCanvasRef.current?.renderPngFrame(frame);
+    };
+    client.addListener(TdpClientEvent.TDP_PNG_FRAME, renderFrame);
+
+    return () => {
+      client.removeListener(TdpClientEvent.TDP_PNG_FRAME, renderFrame);
+    };
+  }, [client, onInitialTdpConnectionSucceeded]);
+
+  useEffect(() => {
+    if (!client) {
+      return;
+    }
+    const renderFrame = (frame: BitmapFrame) => {
+      onInitialTdpConnectionSucceeded();
+      tdpClientCanvasRef.current?.renderBitmapFrame(frame);
+    };
+    client.addListener(TdpClientEvent.TDP_BMP_FRAME, renderFrame);
+
+    return () => {
+      client.removeListener(TdpClientEvent.TDP_BMP_FRAME, renderFrame);
+    };
+  }, [client, onInitialTdpConnectionSucceeded]);
+
+  useEffect(() => {
+    if (!client) {
+      return;
+    }
+    const clear = () => tdpClientCanvasRef.current?.clear();
+    client.addListener(TdpClientEvent.RESET, clear);
+
+    return () => {
+      client.removeListener(TdpClientEvent.RESET, clear);
+    };
+  }, [client]);
+
+  useEffect(() => {
+    if (!client) {
+      return;
+    }
+    const setResolution = (spec: ClientScreenSpec) =>
+      tdpClientCanvasRef.current?.setResolution(spec);
+    client.addListener(TdpClientEvent.TDP_CLIENT_SCREEN_SPEC, setResolution);
+
+    return () => {
+      client.removeListener(
+        TdpClientEvent.TDP_CLIENT_SCREEN_SPEC,
+        setResolution
+      );
+    };
+  }, [client]);
+
   return (
-    <Flex flexDirection="column">
+    <Flex
+      flexDirection="column"
+      css={`
+        // Fill the window.
+        position: absolute;
+        width: 100%;
+        height: 100%;
+      `}
```

Comment:
why is this css needed now? it filled the window previously?

Reply:
Right, but previously it worked in the other way. We were reading the window size, subtracted the top bar, and set the canvas size. Now, we let the container grow as much as possible, the canvas will fill it.
The tricky part is the `position: absolute`, I needed this to keep the container from growing outside the window.